### DISCUSSION TITLE
Best practice for entities with interdependent properties

### DISCUSSION THREAD

**Discussion by armellin** [symfony/symfony#46141] [Q&A]
Hello,

I have a question about Symfony + Doctrine .

Suppose I have an "Animal" Entity, with two properties: "type" and "name".
The possible values ​​of "name" depend on the value of "type", for example: if "type" is "cat" then "name" can be chosen between "Oscar" and "Kitty"; if "type" is "dog" then "name" can be chosen between "Bob" and "Charlie".
An Animal is valid only if both properties are set (not null).
The process of creating a new instance goes through two stages: in the first stage we select the "type" from a ChoiceType field with "cat" and "dog" as options, then a second module will ask you to choose the "name" among the valid ones, according to "type".

One way to implement this is persisting the new instance after selecting just the "type" in the first form, then the second form can retrieve the entity from the database and be populated with the correct choices for "name".
This is perhaps the easiest way, but I don't really like having "invalid" entities in the database, as before submitting the second form.

Any suggestions?

Thank you for your help.

**Comment by norkunas** [comment]
You could use DTO, validate it and then map the data to the entity.

**Comment by MichaelBrauner** [comment]
You can use doctrine inheritance (having an Animal Entity - Than extend it with Cat and Dog). 
Your form could be a polymorphic form working with stimulus. 
It's then like you said - you choose from a dropdown and the form changes to the right one with the correct choices. 

There is the  `https://github.com/infinite-networks/InfiniteFormBundle` 
But I wrote my own Bundle for that. I am not done with testing and documenting it - but if you need that I can realease it and give you an example how to implement it. It's very easy. Something like that: 

```php
 ->add('animal', ParagraphsType::class, [
                'types' => [
                    Cat::class,
                    Dog::class,
                ],
                'allow_add' => true,
                'allow_delete' => true,
                'by_reference' => false,
                'max_items' => 1
            ]);
```

This way you have exactly what u want't in one form (step). And you can write your own custom validator and use it inside your form and/or the entity.
